<a href="https://colab.research.google.com/github/vedashrith/Data-Science/blob/main/Auto_imports_Analysis3_Classification.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## INSTRUCTIONS

Every learner should submit his/her own homework solutions. However, you are allowed to discuss the homework with each other– but everyone must submit his/her own solution; you may not copy someone else’s solution.

The homework consists of two parts:
1.	Data from our life
2.	Classification

Follow the prompts in the attached jupyter notebook. We are using the same data as for the previous homeworks. Use the version you created called df2 where you already cleaned, dropped some of the variables but did not create dummy variables. Instead of creating dummy variables, you have to recode this column as suggested bellow.
Add markdown cells to your analysis to include your solutions, comments, answers. Add as many cells as you need, for easy readability comment when possible.

**Note:** This homework has a bonus question, so the highest mark that can be earned is a 105.
Submission: Send in both a ipynb and a pdf file of your work.
Good luck!



# 1. Data from our lives:

### Describe a situation or problem from your job, everyday life, current events, etc., for which a classification would be appropriate.

##e-mail Spam Filter
In our daily life, we receive hundreds of mails, few of which are promotions, newsletters, spam or mails which we didn't signup for. These mails accumulate in our mail box and waste our storage and time. such mails need to be filtered and deleted because there might be presence of important mails which should not be deleted. In this scenario, classification plays a crucial role in automating the task to filter spam mails and important mails. The features are mail content, sender email address, subject, etc., and the targest variable is binary: spam or not spam. The goal is to trian the ML MODEL TO identify the incoming mails as spam or not spam based on the provided features.

# 2. Preprocessing

In our class we covered multiple classification methods. In this part of the home work you can compare them

**Use the dataset 'auto_imports1.csv' from our previous homeworks. More specifically, use the version you created called df2 where you already cleaned, dropped some of the variables but DID NOT CREATE dummy variables. Follow the prompts to complete the homework.**

In [ ]:
import pandas as pd

df = pd.read_csv('df3.csv')

In [ ]:
df.head()

fuel_type  wheel_base  length  width  heights  curb_weight  engine_size  \
0       gas        88.6   168.8   64.1     48.8         2548          130   
1       gas        88.6   168.8   64.1     48.8         2548          130   
2       gas        94.5   171.2   65.5     52.4         2823          152   
3       gas        99.8   176.6   66.2     54.3         2337          109   
4       gas        99.4   176.6   66.4     54.3         2824          136   

   bore  stroke  comprassion  horse_power  peak_rpm  city_mpg  highway_mpg  \
0  3.47    2.68          9.0        111.0    5000.0        21           27   
1  3.47    2.68          9.0        111.0    5000.0        21           27   
2  2.68    3.47          9.0        154.0    5000.0        19           26   
3  3.19    3.40         10.0        102.0    5500.0        24           30   
4  3.19    3.40          8.0        115.0    5500.0        18           22   

   price  
0  13495  
1  16500  
2  16500  
3  13950  
4  17450

The dataset is cleaned as per requirement and saved as 'df3.csv' and imported here to perform classification.

## 2.1 **Replace ['gas', 'diesel'] string values to [0, 1]**

In [ ]:
#Your code
df['fuel_type'] = df['fuel_type'].replace({'gas':0, 'diesel':1})

Replaced the variable 'fuel_type' content from gas and diesel to 0 and 1.

In [ ]:
df.head()

fuel_type  wheel_base  length  width  heights  curb_weight  engine_size  \
0          0        88.6   168.8   64.1     48.8         2548          130   
1          0        88.6   168.8   64.1     48.8         2548          130   
2          0        94.5   171.2   65.5     52.4         2823          152   
3          0        99.8   176.6   66.2     54.3         2337          109   
4          0        99.4   176.6   66.4     54.3         2824          136   

   bore  stroke  comprassion  horse_power  peak_rpm  city_mpg  highway_mpg  \
0  3.47    2.68          9.0        111.0    5000.0        21           27   
1  3.47    2.68          9.0        111.0    5000.0        21           27   
2  2.68    3.47          9.0        154.0    5000.0        19           26   
3  3.19    3.40         10.0        102.0    5500.0        24           30   
4  3.19    3.40          8.0        115.0    5500.0        18           22   

   price  
0  13495  
1  16500  
2  16500  
3  13950  
4  17450

## 2.2 : Define your X and y: your dependent variable is fuel_type, the rest of the variables are your independent variables

In [ ]:
#your code
y = df['fuel_type']
X = df.drop(columns = ['fuel_type'])

The variable y is the Target variable.
The variable X consists of all the predictor variables/features.

## 2.3 Split your data into training and testing set. Use test_size=0.3, random_state=746 !

In [ ]:
#your code
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X , y, test_size = 0.3, random_state = 746)

We have imported train_test_split from sklearn.model_selection package to split the data into train and test datasets. The train dataset has 70% of data and test has 30% of data.

# 3. Classification

### 3.1 Use Logistic regression to classify your data. Print/report your confusion matrix, classification report and AUC

In [ ]:
#your code
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import confusion_matrix, classification_report, roc_auc_score

log = LogisticRegression(max_iter=1000)
log.fit(X_train, y_train)

y_pred = log.predict(X_test)

print("Confusion matrix:\n",confusion_matrix(y_test, y_pred))

print("Classification report:\n",classification_report(y_test, y_pred))
print("AUC score:\n", roc_auc_score(y_test, y_pred))

Confusion matrix:
 [[50  0]
 [ 0  9]]
Classification report:
               precision    recall  f1-score   support

           0       1.00      1.00      1.00        50
           1       1.00      1.00      1.00         9

    accuracy                           1.00        59
   macro avg       1.00      1.00      1.00        59
weighted avg       1.00      1.00      1.00        59

AUC score:
 1.0


Logistic Regression:
Logistic function uses a logistic function to transform the output of the linear function into a probabity value.

We train the model by fitting X_train and y_train and predict the target variable as y_pred.

The output of this model indicates excellent performance by the model. The Confusion matrix shows no false positives and no false negatives, that means the model predicted all the instances correctly.

The Classification report gives the perfect precision, recall and f1 scores for both the classes: gas and diesel.

The AUC score of 1.0 indicates the model is able to distinguish between positive and negative classes perfectly.

The overall performance of this Logistic Regression model is high and perfect.

### 3.2 Use Naive Bayes to classify your data. Print/report your confusion matrix, classification report and AUC

In [ ]:
#your code
from sklearn.naive_bayes import GaussianNB

nb = GaussianNB()
nb.fit(X_train, y_train)

y_pred = nb.predict(X_test)
y_pred_prob = nb.predict_proba(X_test)[:,1]
auc = roc_auc_score(y_test, y_pred_prob)

print('Confusion matrix:\n',confusion_matrix(y_test, y_pred))
print('Classification Report:\n', classification_report(y_test, y_pred))
print('AUC score:',auc)

Confusion matrix:
 [[50  0]
 [ 0  9]]
Classification Report:
               precision    recall  f1-score   support

           0       1.00      1.00      1.00        50
           1       1.00      1.00      1.00         9

    accuracy                           1.00        59
   macro avg       1.00      1.00      1.00        59
weighted avg       1.00      1.00      1.00        59

AUC score: 1.0


Naive Bayes Classification Technicque is extremely fast relative to other classification algorithms. It works on Bayes' Theorem of probability to predict the class of unknown datasets.

We have imported GaussianNB from sklearn.naive_bayes library

We train the model by fitting X_train and y_train and predict the target variable as y_pred.

The output of this model indicates excellent performance by the model. The Confusion matrix shows no false positives and no false negatives, that means the model predicted all the instances correctly.

The Classification report gives the perfect precision, recall and f1 scores for both the classes: gas and diesel.

The AUC score of 1.0 indicates the model is able to distinguish between positive and negative classes perfectly.

The overall performance of this model is high and perfect.



### 3.3 Use KNN to classify your data. First find the optimal k and than run you classification. Print/report your confusion matrix, classification report and AUC

In [ ]:
#your code
from sklearn.model_selection import GridSearchCV
from sklearn.neighbors import KNeighborsClassifier

par = {'n_neighbors': range(1, 20)}
knn = KNeighborsClassifier()
g_search = GridSearchCV(knn, par, cv = 5, scoring = 'accuracy')
g_search.fit(X_train, y_train)

k = g_search.best_params_['n_neighbors']
print("optimal k value:", k)

opt_knn = KNeighborsClassifier(n_neighbors = k)
opt_knn.fit(X_train, y_train)
y_pred = opt_knn.predict(X_test)

print("Confusion Matrix:\n",confusion_matrix(y_test, y_pred))
print("Classification Report:\n", classification_report(y_test, y_pred))
print("AUC score", roc_auc_score(y_test, y_pred))

optimal k value: 2
Confusion Matrix:
 [[50  0]
 [ 8  1]]
Classification Report:
               precision    recall  f1-score   support

           0       0.86      1.00      0.93        50
           1       1.00      0.11      0.20         9

    accuracy                           0.86        59
   macro avg       0.93      0.56      0.56        59
weighted avg       0.88      0.86      0.82        59

AUC score 0.5555555555555556


KNN: K-Nearest Neighbors (KNN) is a straightforward and intuitive machine learning algorithm used for both classification and regression tasks. Given a new, unseen data point, KNN identifies its nearest neighbors based on a similarity measure. It then classifies the new data point by a majority vote among its K-nearest neighbors. The class assigned to the new data point is the most common class among its K-nearest neighbors.  

Optimal K value is calculated as 2.  

The number of False negatives are 8 that indicates the model is facing issues with the fuel_type: 1. The precision and recall for diesel class are significantly affected. This is due to imbalanced nature of the dataset. The AUC score is 0.55 which indicates that the model performed poorly.

### 3.4 Choose one: SVM or Random Forest to classify your data. Print/report your confusion matrix, classification report and AUC

In [ ]:
#your code
from sklearn.ensemble import RandomForestClassifier

rf = RandomForestClassifier()
rf.fit(X_train, y_train)

y_pred = rf.predict(X_test)
y_pred_prob = rf.predict_proba(X_test)[:,1]
print("Confusion Matrix:\n",confusion_matrix(y_test, y_pred))
print("Classification Report:\n",classification_report(y_test, y_pred))
print("AUC score:", roc_auc_score(y_pred, y_pred_prob))

Confusion Matrix:
 [[50  0]
 [ 0  9]]
Classification Report:
               precision    recall  f1-score   support

           0       1.00      1.00      1.00        50
           1       1.00      1.00      1.00         9

    accuracy                           1.00        59
   macro avg       1.00      1.00      1.00        59
weighted avg       1.00      1.00      1.00        59

AUC score: 1.0


Random Forest is a versatile machine learning method capable of performing both regression and classification tasks.  

We have imported RandomForestClassifier from sklearn.ensemble library

We train the model by fitting X_train and y_train and predict the target variable as y_pred.

The output of this model indicates excellent performance by the model. The Confusion matrix shows no false positives and no false negatives, that means the model predicted all the instances correctly.

The Classification report gives the perfect precision, recall and f1 scores for both the classes: gas and diesel wit 100% accuracy.

The AUC score of 1.0 indicates the model is able to distinguish between positive and negative classes perfectly.

The overall performance of this model is high and perfect.


### 3.5 Compare your results and comment on your findings. Which one(s) did the best job? What could have been the problem with the ones that did not work? etc.

The Confusion Matrix, Classification Report and AUC score of differnt models are reported above. The models: Logistic Regression, Naive Bayes Classifier and Random Forest Classifier performed exceptionally well. There are False positives and False negatives which these models predicted correctly. These models have acheived 100% accuracy, Recall, Precision and F1 scores. This proves that all these models made no mistakes in classifying the test data set.  

The KNN model performed poorly with an AUC score of 0.55. All other models performs 45% better than KNN model. This model has 8 False negatives whereas all other models have 0. The accuracy is 86%. The Recall and F1 score are affected as well. This could be due to the imbalanced nature of the dataset. Hence, it is recommended to perform dataset balancing techniques on the dataset and test the model.

## 4. Bonus question (5 extra points)
**Try to fix the inbalanced nature of the data with a tool from the lecture. Run one of the classification methods (preferable one that "failed" before) and see if you get better results.**

In [ ]:
#your code
from imblearn.over_sampling import SMOTE

smote = SMOTE(random_state=746)
X_balanced, y_balanced = smote.fit_resample(X_train, y_train)

opt_knn = KNeighborsClassifier(n_neighbors=k)
opt_knn.fit(X_balanced, y_balanced)
y_pred = opt_knn.predict(X_test)

print("Confusion Matrix:\n", confusion_matrix(y_test, y_pred))
print("Classification Report:\n", classification_report(y_test, y_pred))
print("AUC score:", roc_auc_score(y_test, y_pred))


Confusion Matrix:
 [[44  6]
 [ 4  5]]
Classification Report:
               precision    recall  f1-score   support

           0       0.92      0.88      0.90        50
           1       0.45      0.56      0.50         9

    accuracy                           0.83        59
   macro avg       0.69      0.72      0.70        59
weighted avg       0.85      0.83      0.84        59

AUC score: 0.7177777777777777


SMOTE (Synthetic Minority Oversampling Technique) is one of the resampling techniques to fix the imbalanced dataset.  
This technique generates synthetic samples from class 1 to balance the class distribution.

The resampled data is stored in X_balanced and y_balanced and fitted in the KNN model.  
There are significant improvements in the Confusion Matrix: The False negative are reduced to 4. In the Classification report, the recall and f1 score have increased. The AUC score has also increased to 72% which indicates improvements in the overall performance model. After balancing the data, the models performs better than before.

